### PR TITLE
Return a mixture of exact and sampled weights as appropriate

### LINKED ISSUE #199: Return a mixture of exact weights and samples in `generate_qpd_samples` when some (but not all) weights are above the threshold implied by `1 / num_samples`
#190 introduced exact weights, but it's currently all or nothing.  Ideally, exact weights will be used for weights above the `1 / num_samples` threshold, and all other weights will be sampled.

### PR DESCRIPTION
**Pull request by garrison**
Fixes #199

Remaining action items:
- [x] update docstring
- [x] rename to `generate_qpd_weights`
- [x] add release note
- [x] do everything I can to make this code more understandable (not sure if this will ever truly be "done", but the code is ready for comments :slightly_smiling_face:)
- [x] update tutorials and "exact weights" how-to guide to be consistent with new method for taking exact weights

### CODE REVIEW

File: circuit_knitting/cutting/qpd/qpd.py [hunk 1]
```diff
@@ -61,66 +73,398 @@ class WeightType(Enum):
...
+
+    return retval
+
+
+def _populate_samples(
+    random_samples: dict[tuple[int, ...], int],
+    num_desired: int,
+    independent_probabilities: Sequence,
+    conditional_probabilities: dict[tuple[int, ...], npt.NDArray[np.float64]],
+    runner: tuple[int, ...] = (),
```
Comment (garrison) [inline circuit_knitting/cutting/qpd/qpd.py:426]: Rename this `running_state`.

**Review by caleb-johnson — COMMENTED** [review]
Looking great, made a few comments and will come back to this for another pass in a bit.

File: releasenotes/notes/generate-qpd-weights-mixture-79c1c420105c8564.yaml [hunk 2]
```diff
@@ -0,0 +1,15 @@
+---
+features:
+  - |
+    :func:`~circuit_knitting.cutting.qpd.generate_qpd_weights` now
```
[the remark this answers is not in the document]
Comment (garrison) [reply releasenotes/notes/generate-qpd-weights-mixture-79c1c420105c8564.yaml:4]: Thanks, I was thinking something similar.  I clarified this in 2e7d7b7465ad7c7ea838782288252b720cf19334.

File: circuit_knitting/cutting/qpd/qpd.py [hunk 3]
```diff
@@ -61,66 +73,398 @@ class WeightType(Enum):
...
+        logger.info("No exact weights")
+    else:
+        logger.info("Some exact weights")
+        for (
+            map_ids,
+            probability,
+        ) in _generate_exact_weights_and_conditional_probabilities(
+            independent_probabilities, threshold
+        ):
+            if len(map_ids) == len(independent_probabilities):
```
[the remark this answers is not in the document]
Comment (garrison) [reply circuit_knitting/cutting/qpd/qpd.py:356]: The reason for this `if`/`else` statement is given by the two different possible types of information the generator can return.  I will add a comment here to make this more clear.

https://github.com/Qiskit-Extensions/circuit-knitting-toolbox/blob/2e7d7b7465ad7c7ea838782288252b720cf19334/circuit_knitting/cutting/qpd/qpd.py#L108-L118

[the remark this answers is not in the document]
Comment (garrison) [reply circuit_knitting/cutting/qpd/qpd.py:361]: Clarified in 44951d8db2f652b93e59000cced60c712e545633.

File: circuit_knitting/cutting/qpd/qpd.py [hunk 4]
```diff
@@ -61,66 +73,398 @@ class WeightType(Enum):
...
+
+
+def generate_qpd_weights(
+    qpd_bases: Sequence[QPDBasis], num_samples: float = 1000
+) -> dict[tuple[int, ...], tuple[float, WeightType]]:
+    """
+    Generate weights from the joint quasiprobability distribution.
+
+    All weights above ``1 / num_samples`` will be returned exactly, while the
+    remaining weights will be sampled from.
```
[the remark this answers is not in the document]
Comment (garrison) [reply circuit_knitting/cutting/qpd/qpd.py]: I rephrased this in 8830c941822cbbf14084f99c74c752ceafc3df4c.

File: circuit_knitting/cutting/qpd/qpd.py [hunk 5]
```diff
@@ -61,66 +73,398 @@ class WeightType(Enum):
...
+
     Args:
-        qpd_bases: The :class:`QPDBasis` objects from which to sample
-        num_samples: Number of random samples to generate
+        qpd_bases: The :class:`QPDBasis` objects from which to generate weights
+        num_samples: Maximum number of weights to generate
```
[the remark this answers is not in the document]
Comment (garrison) [reply circuit_knitting/cutting/qpd/qpd.py]: I also rephrased this a bit in 8830c941822cbbf14084f99c74c752ceafc3df4c.

**Review by caleb-johnson — APPROVED** [review]
Looks great. A couple of pedantic suggestions, but LGTM 👍

File: docs/circuit_cutting/tutorials/02_gate_cutting_to_reduce_circuit_depth.ipynb [hunk 6]
```diff
@@ -291,9 +291,11 @@
...
+    "The number of weights taken from the distribution is controlled by `num_samples`. Each weight whose absolute value is above a threshold of 1 / `num_samples` will be evaluated exactly.  The remaining weights -- those in the tail of the distribution -- will then be sampled from, resulting in at most `num_samples` unique weights.\n",
+    "\n",
+    "Much of the circuit cutting literature describes a process where we sample from the distribution, take a single shot, then sample from the distribution again and repeat; however, this is not feasible in practice. The total number of shots needed grows exponentially with the number of cuts, and single shot experiments via Qiskit Runtime quickly become untenable. Instead, we take an equivalent number of shots for each considered subexperiment, and send them to the backend(s) in batches. During reconstruction, each subexperiment contributes to the final result with proportion equal to its weight.  We just need to ensure the number of shots we take is appropriate for the heaviest weights, and thus, appropriate for all weights."
```
Comment (caleb-johnson) [inline docs/circuit_cutting/tutorials/02_gate_cutting_to_reduce_circuit_depth.ipynb:298]:
```suggestion
    "Much of the circuit cutting literature describes a process where we sample from the distribution, take a single shot, then sample from the distribution again and repeat; however, this is not feasible in practice. The total number of shots needed grows exponentially with the number of cuts, and single shot experiments via Qiskit Runtime quickly become untenable. Instead, we take an equivalent number of shots for each considered subexperiment and send them to the backend(s) in batches. During reconstruction, each subexperiment contributes to the final result with proportion equal to its weight.  We just need to ensure the number of shots we take is appropriate for the heaviest weights, and thus, appropriate for all weights."
```

Comment (caleb-johnson) [reply docs/circuit_cutting/tutorials/02_gate_cutting_to_reduce_circuit_depth.ipynb:298]: I never get to correct any grammar errors in your docs. Please indulge me here lol

Comment (caleb-johnson) [reply docs/circuit_cutting/tutorials/02_gate_cutting_to_reduce_circuit_depth.ipynb:298]: I'm not even convinced your comma is wrong. The second thought could be an imperative sentence, but I think not ;)

File: docs/circuit_cutting/tutorials/02_gate_cutting_to_reduce_circuit_depth.ipynb [hunk 7]
```diff
@@ -291,9 +291,11 @@
...
     "\n",
-    "To simulate the expectation value of the full-sized circuit, many subexperiments are generated from the decomposed gates' joint quasiprobability distribution and then executed on one or more backends. The number of samples taken from the distribution is controlled by `num_samples`. If the smallest sample probability in the joint quasi-probability distribution is above 1 / `num_samples`, exact weights will be used.  Otherwise, the weights will be sampled, resulting in at most `num_samples` unique samples.  In the future, this function will be changed to return a mixture of exact and sampled weights, so that the largest weights will be calculated exactly, while weights in the tail of the distribution will be sampled."
+    "The number of weights taken from the distribution is controlled by `num_samples`. Each weight whose absolute value is above a threshold of 1 / `num_samples` will be evaluated exactly.  The remaining weights -- those in the tail of the distribution -- will then be sampled from, resulting in at most `num_samples` unique weights.\n",
```
[the remark this answers is not in the document]
Comment (caleb-johnson) [reply docs/circuit_cutting/tutorials/02_gate_cutting_to_reduce_circuit_depth.ipynb:296]:
```suggestion
    "The number of weights taken from the distribution is controlled by `num_samples`. Each weight whose absolute value is above a threshold of 1 / `num_samples` will be evaluated exactly.  The remaining low-probability elements -- those in the tail of the distribution -- will then be sampled from, resulting in at most `num_samples` unique weights.\n",
```

**Pull request merged**